lib/common/pid.c

```diff
@@ -20,65 +20,71 @@
 #include <crm/crm.h>
 
 int
-crm_pid_active(long pid, const char *daemon)
+pcmk__pid_active(pid_t pid, const char *daemon)
 {
-    static int last_asked_pid = 0;  /* log spam prevention */
+    static pid_t last_asked_pid = 0;  /* log spam prevention */
 #if SUPPORT_PROCFS
     static int have_proc_pid = 0;
 #else
     static int have_proc_pid = -1;
 #endif
     int rc = 0;
+    bool no_name_check = ((daemon == NULL) || (have_proc_pid == -1));
 
     if (have_proc_pid == 0) {
         /* evaluation of /proc/PID/exe applicability via self-introspection */
         char proc_path[PATH_MAX], exe_path[PATH_MAX];
-        snprintf(proc_path, sizeof(proc_path), "/proc/%lu/exe",
-                 (long unsigned int) getpid());
+        snprintf(proc_path, sizeof(proc_path), "/proc/%lld/exe",
+                 (long long) getpid());
         have_proc_pid = 1;
         if (readlink(proc_path, exe_path, sizeof(exe_path) - 1) < 0) {
             have_proc_pid = -1;
         }
     }
 
     if (pid <= 0) {
-        return -1;
+        return EINVAL;
+    }
 
-    } else if ((rc = kill(pid, 0)) < 0 && errno == ESRCH) {
-        return 0;  /* no such PID detected */
+    rc = kill(pid, 0);
+    if ((rc < 0) && (errno == ESRCH)) {
+        return ESRCH;  /* no such PID detected */
 
-    } else if (rc < 0 && (daemon == NULL || have_proc_pid == -1)) {
+    } else if ((rc < 0) && no_name_check) {
+        rc = errno;
         if (last_asked_pid != pid) {
-            crm_info("Cannot examine PID %ld: %s", pid, strerror(errno));
+            crm_info("Cannot examine PID %lld: %s",
+                     (long long) pid, strerror(errno));
             last_asked_pid = pid;
         }
-        return -2;  /* errno != ESRCH */
+        return rc; /* errno != ESRCH */
 
-    } else if (rc == 0 && (daemon == NULL || have_proc_pid == -1)) {
-        return 1;  /* kill as the only indicator, cannot double check */
+    } else if ((rc == 0) && no_name_check) {
+        return pcmk_rc_ok; /* kill as the only indicator, cannot double check */
 
     } else if (daemon != NULL) {
         /* make sure PID hasn't been reused by another process
            XXX: might still be just a zombie, which could confuse decisions */
         bool checked_through_kill = (rc == 0);
         char proc_path[PATH_MAX], exe_path[PATH_MAX], myexe_path[PATH_MAX];
-        snprintf(proc_path, sizeof(proc_path), "/proc/%ld/exe", pid);
+        snprintf(proc_path, sizeof(proc_path), "/proc/%lld/exe",
+                 (long long) pid);
 
         rc = readlink(proc_path, exe_path, sizeof(exe_path) - 1);
-        if ((rc < 0) && (errno == EACCES)) {
+        if (rc < 0) {
             if (last_asked_pid != pid) {
-                crm_info("Could not read from %s: %s", proc_path,
-                         strerror(errno));
+                crm_err("Could not read from %s: %s " CRM_XS " errno=%d",
+                        proc_path, strerror(errno), errno);
```

Comment:
I'm seeing this error with SUSE Linux:

`pacemaker-controld  [29351] (pcmk__pid_active@pid.c:77)   error: Could not read from /proc/18493/exe: Permission denied | errno=13`

`18493` is the "sbd: inquisitor" daemon. I don't see any permission limits preventing `hacluster` user from accessing the file  just according to the permission modes though.

Not sure if there's any security policy hidden somewhere. It probably has been like this but never got noticed with the previous info logging level ... I'm wondering how it works for you with non-SUSE systems.